hips/hip-0004.md

```diff
@@ -0,0 +1,114 @@
+type: "process"
+status: "draft"
+---
+
+## Abstract
+
+Define and document the backwards-compatibility rules that apply to Helm minor and patch releases.
+
+## Motivation
+
+Helm aims to strictly respect [Semantic Versioning][semver], where minor and patch releases must be 100% backwards compatible.
+Backwards compatibility implies that when moving to a new backwards-compatible release:
+* tools (such as Continuous Integration tools) using the helm CLI will not require any changes
+* programs using the Helm Go API will continue to compile 
+
+Some contributors to the Helm code base may not be comfortable with the details of what constitutes a compatibility-breaking change.
+Furthermore, in some cases, there can be a difference of opinion, even amongst the maintainers, on what constitutes such a change.
+
+Having a well-defined, documented set of backwards-compatibility rules for the Helm code-base will allow to:
+* avoid compatibility-breaking changes being merged by mistake
+* guide contributors in understanding helm's backwards-compatibility requirements
+* remove any uncertainty when it comes to enforcing backwards compatibility
+* provide a basis for discussion and possible updates to how helm handles backwards compatibility
+
+## Rationale
+
+The maintainers have agreed that a formal definition of Helm's backwards-compatibility rules would be beneficial.
+
+## Specification
+
+The following compatibility rules will apply to helm minor and patch releases:
+* exported Go APIs will remain compatible as per this [article][go-module-comp]
+* CLI commands and flags:
+    * commands and flags are case-sensitive
+    * commands and flags must not be removed
+    * commands and flags must not be renamed
+    * commands and flags must not be moved elsewhere in the command hierarchy
+    * commands and flags cannot be "repurposed" to provide a different behavior than the original (even if deprecated)
+    * flags must not change type.  Exceptions are when the new type is a superset of the original type (e.g., int8 to int32, int to float, float to string)
+    * these rules apply as much to the short name as to the long name of a flag, if present.  Specifically, a short name cannot be changed in any way even if its long name is kept the same, and vice versa
+    
+* CLI output:
+    * the help text may change
+    * the format of structured output (tables, JSON, YAML) must not change
+    * errors can change if doing so increases their usability
+    * a message containing a "bug" (including spelling errors, misinformation, or egregious grammar errors), can be fixed
+    * return codes should not be changed unless they are incorrect
+
+* File formats (such as index.yaml, values.yaml, Chart.yaml):
+    * No fields may be removed, added, or modified on the Chart.yaml file
+    * No fields may be modified or removed on...
+        * index.yaml
+        * repositories.yaml
+    * Any added fields must be accompanied by backward compatibility checks
```

Comment:
I'm not sure about this part. _Chart.yaml_ can't have new fields added to it but _index.yaml_ can? If the reason for this is the strict parsing than both of these cases should not have new fields added. And, possibly a note explaining the strict parser issue.

Reply:
So, should adding fields be allowed in every case but only as optional?  
Why couldn't we add an optional field to Chart.yaml?

Reply:
The problem here is that the security best-practices that were recommended to us required that we use _strict_ parsing on YAML docs, which entails throwing an error any time it hits an unknown field.

But that means that we couldn't add purely optional fields. An existing parser would fail if it hit those fields.

We got around this in a few places by defining a particular field as a map[string]string. But net-new fields would still have this problem

Reply:
So IIUC, no new fields can be added to any of these yaml file due to the strict parsing?  I can add an explanation in the HIP.

I would have thought that strict parsing should be applied to yaml files that helm receives such as Chart.yaml and index.yaml, but not to yaml files helm generates like repositories.yaml?

Reply:
In the last dev call, we brainstormed a bit on this and determined that adding new optional fields in a newer version of Helm would mean that older versions with no knowledge of said field being optional could result in a breaking change.

As in, if you have a structure like this in Helm 3.x:

```
type Foo struct {
    Bar string
}
```

and then in Helm 3.x+1, you add an optional field:

```
type Foo struct {
    Bar string
    Car string `json:",omitempty"`
```

With `{ "bar": "hello" }` as input, both versions would parse it successfully. But with `{"bar": "hello", "car": "goodbye"}` Helm 3.x would return an error.

We couldn't find a way to resolve this without writing a new parser that disallows duplicate keys while allowing unknown keys. Because the original issue was to fix the following case from happening:

```
car: "hello"
car: "goodbye"
```

in other words I'm not sure we can stipulate additions/removals on ANY file until we address this limitation. :'(

Reply:
I think we are ready to cal this one done. As @bacongobbler notes, a parser issue is the reason why index.yaml (and repositories.yaml) are considered locked down. Chart.yaml is locked for compatibility reasons.

Files that are locked (cannot have ad-hoc fields):

- Chart.yaml
- index.yaml
- repositories.yaml

These are all correctly called out in the PR.

So I think we can close and merge this PR.

Reply:
The HIP currently states
```
* File formats (such as index.yaml, values.yaml, Chart.yaml):
    * No fields may be removed, added, or modified on the Chart.yaml file
    * No fields may be modified or removed in:
        * index.yaml
        * repositories.yaml
```
but IIUC from @technosophos previous comment, both `index.yaml` and `repositories.yaml` should also not have any fields added to them so that they keep working with older versions of helm.  So I believe I should update the HIP to say:
```
* File formats (such as index.yaml, values.yaml, Chart.yaml):
    * No fields may be modified or removed
    * No fields may be added in:
        * Chart.yaml
        * index.yaml
        * repositories.yaml
```
Can someone confirm this is correct to finish off this HIP?